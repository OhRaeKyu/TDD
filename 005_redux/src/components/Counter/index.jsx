import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { increase, decrease } from '../../modules/counterModule';

export default React.memo(function Counter() {
  const dispatch = useDispatch();
  const { number } = useSelector((state) => state.counter);

  const handleIncrease = useCallback(() => {
    dispatch(increase());
  }, [dispatch]);

  const handleDecrease = useCallback(() => {
    dispatch(decrease());
  }, [dispatch]);

  return (
    <div>
      <p>{number}</p>
      <button type="button" onClick={handleIncrease}>
        +
      </button>
      <button type="button" onClick={handleDecrease}>
        -
      </button>
    </div>
  );
});
