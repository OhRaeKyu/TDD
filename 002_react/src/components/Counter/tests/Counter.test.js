import { render, screen, fireEvent } from '@testing-library/react';
import Counter from '../index';

describe('Counter Component', () => {
  it('matches snapshot', () => {
    const utils = render(<Counter />);
    expect(utils.container).toMatchSnapshot();
  });

  it('has a number and two buttons', () => {
    render(<Counter />);

    const number = screen.getByText('0');
    const increaseBtn = screen.getByText('+');
    const decreaseBtn = screen.getByText('-');

    expect(number).toBeInTheDocument();
    expect(increaseBtn).toBeInTheDocument();
    expect(decreaseBtn).toBeInTheDocument();
  });

  it('increase', () => {
    render(<Counter />);

    const number = screen.getByText('0');
    const increaseBtn = screen.getByText('+');

    fireEvent.click(increaseBtn);
    expect(number).toHaveTextContent('1');
    fireEvent.click(increaseBtn);
    fireEvent.click(increaseBtn);
    expect(number).toHaveTextContent('3');
  });

  it('decrease', () => {
    render(<Counter />);

    const number = screen.getByText('0');
    const decreaseBtn = screen.getByText('-');

    fireEvent.click(decreaseBtn);
    expect(number).toHaveTextContent('-1');
    fireEvent.click(decreaseBtn);
    fireEvent.click(decreaseBtn);
    expect(number).toHaveTextContent('-3');
  });
});
