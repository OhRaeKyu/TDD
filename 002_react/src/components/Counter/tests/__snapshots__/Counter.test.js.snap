// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component matches snapshot 1`] = `
<div>
  <div>
    <h2>
      0
    </h2>
    <button>
      +
    </button>
    <button>
      -
    </button>
  </div>
</div>
`;
